src/ServiceStack/DependencyInjection/DependencyService.cs

```diff
@@ -37,22 +39,82 @@ public DependencyService()
             _containerBuilder = new ContainerBuilder();
         }
 
-        public void RegisterTypeAsItself(Type classType, Sharing sharing = Sharing.None)
+        public void Register(Type implementingType, Sharing sharing, bool registerAsImplementedInterfaces, bool includeNonPublicConstructors)
         {
-            var registration = _containerBuilder.RegisterType(classType);
-            SetSharing(registration, sharing);
+            if (implementingType.IsGenericType)
+            {
+                var registration = _containerBuilder.RegisterGeneric(implementingType);
+                if (registerAsImplementedInterfaces)
+                {
+                    registration = registration.AsImplementedInterfaces();
+                }
+                if (includeNonPublicConstructors)
+                {
+                    registration = registration.FindConstructorsWith(type => type.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic));
+                }
+                registration = SetRegistrationLifetime(registration, sharing);
+            }
+            else
+            {
+                var registration = _containerBuilder.RegisterType(implementingType).AsSelf();
+                if (registerAsImplementedInterfaces)
+                {
+                    registration = registration.AsImplementedInterfaces();
+                }
+                if (includeNonPublicConstructors)
+                {
+                    registration = registration.FindConstructorsWith(type => type.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic));
+                }
+                registration = SetRegistrationLifetime(registration, sharing);
```

Comment:
I would like to, but sadly the Autofac response objects to `RegisterGeneric()` and `RegisterType()` are different.  In some cases, you can share the code that uses them (see the implementation of `SetRegistrationLifetime()`), but for the rest of this, you can't.  If you can figure out a way to write C# that will work, let me know.